### PR TITLE
NxForm RSC-374

### PR DESCRIPTION
**Pull request by rpokorny**
https://issues.sonatype.org/browse/RSC-374

This PR is a little code-heavy, so I'm requesting review from outside of just @sonatype/team-plaid .

### CODE REVIEW

File: lib/src/base-styles/_nx-forms.scss [hunk 1]
```diff
@@ -6,6 +6,10 @@
  */
 @import '../scss-shared/nx-variables';
 
+.nx-form {
+  position: relative;
+}
```
Comment (rpokorny) [inline lib/src/base-styles/_nx-forms.scss:11]: For non-fullscreen submit masks

File: lib/src/components/NxTextInput/NxTextInput.tsx [hunk 2]
```diff
@@ -58,8 +58,8 @@ const NxTextInput = forwardRef<TextInputElement, Props>(
           firstValidationError = validatable && getFirstValidationError(validationErrors),
           internalClassName = classnames('nx-text-input', className, {
             pristine: isPristine,
-            invalid: validatable && isInvalid,
-            valid: validatable && !isInvalid,
+            invalid: !isPristine && validatable && isInvalid,
+            valid: !isPristine && validatable && !isInvalid,
```
Comment (rpokorny) [inline lib/src/components/NxTextInput/NxTextInput.tsx:62]: In an `NxForm`, you'll typically want to compute validation errors even for pristine inputs, as they should affect the disabled state of the Submit button (e.g. if there are required fields the button should initially be disabled until they are filled in).  The changes here and the `stateHelpers` make that much more convenient.

File: lib/src/util/validationUtil.ts [hunk 3]
```diff
@@ -42,3 +43,7 @@ export function getFirstValidationError(validationErrors: ValidationErrors | und
     return validationErrors;
   }
 };
+
+export function combineValidationErrors(...validationErrors: (ValidationErrors | undefined)[]): ValidationErrors {
```
Comment (rpokorny) [inline lib/src/util/validationUtil.ts:47]: Realized after posting the PR that this needs docs. I'll add them before this is merged.  TODO

**Review by Blacktiger — APPROVED** [review]
Don't forget your TODO, otherwise looks good to me with only a couple of minor comments. I'm pretty sure this will work well with the way the nxrm codebase currently works.

File: lib/src/util/validationUtil.ts (hunk 3 above)
Comment (rpokorny) [reply lib/src/util/validationUtil.ts:47]: > Realized after posting the PR that this needs docs. I'll add them before this is merged. TODO

d44ef6da

File: lib/src/components/NxForm/NxForm.tsx [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+          onSubmitProp();
+        }
+      }
+
+      const renderForm = () => (
+        <form ref={ref} className={formClasses} onSubmit={onSubmit} { ...formAttrs }>
+          { getChildren() }
+          <footer className="nx-footer">
+            { submitError &&
+              <NxLoadError titleMessage="An error occurred saving data."
```
[the remark this answers is not in the document]
Comment (rpokorny) [reply lib/src/components/NxForm/NxForm.tsx]: > I wonder if we might want to customize this message?

3fb58d12

File: lib/src/components/NxTextInput/__tests__/NxTextInput.test.tsx [hunk 5]
```diff
@@ -27,7 +27,7 @@ describe('NxTextInput', function() {
     expect(getShallowComponent()).toContainMatchingElement('.nx-text-input__box .nx-icon--invalid');
   });
 
-  it('contains the first validation message', function() {
+  it('contains the first validation message if isPristine is not trie', function() {
```
[the remark this answers is not in the document]
Comment (rpokorny) [reply lib/src/components/NxTextInput/__tests__/NxTextInput.test.tsx]: > spelling?

e01c4e3c

**Review by glenhunter — APPROVED** [review]
This is really cool, I'll leave it to @Blacktiger to review the more sophisticated parts, but what I understand looks great.

**Pull request merged**